File: src/lang/csharp/Chamber.cs
namespace CaveExplorer;

public class Chamber
{
    public string Name { get; }
    public double WidthMeters { get; }
    public double HeightMeters { get; }
    public double LengthMeters { get; }
    public double DepthMeters { get; }
    public double Temperature { get; set; }
    public double Humidity { get; set; }
    public List<GeologicalFormation> Formations { get; } = new();

    public Chamber(string name, double width, double height, double length, double depth)
    {
        Name = name;
        WidthMeters = width;
        HeightMeters = height;
        LengthMeters = length;
        DepthMeters = depth;
        Temperature = 12.0; // Default cave temperature in Celsius
        Humidity = 95.0;    // Default cave humidity percentage
    }

    public double Volume => WidthMeters * HeightMeters * LengthMeters;

    public void AddFormation(GeologicalFormation formation)
    {
        Formations.Add(formation);
    }

    public IEnumerable<GeologicalFormation> GetFormationsByType(FormationType type) =>
        Formations.Where(f => f.Type == type);

    public override string ToString() =>
        $"Chamber[{Name}, {Volume:F1}m³, depth={DepthMeters}m, {Formations.Count} formations]";
}

File: src/lang/java/Chamber.java
package caves;

import java.util.ArrayList;
import java.util.List;

public class Chamber {
    private final String name;
    private final double widthMeters;
    private final double heightMeters;
    private final double lengthMeters;
    private final List<Formation> formations;

    public Chamber(String name, double widthMeters, double heightMeters, double lengthMeters) {
        this.name = name;
        this.widthMeters = widthMeters;
        this.heightMeters = heightMeters;
        this.lengthMeters = lengthMeters;
        this.formations = new ArrayList<>();
    }

    public void addFormation(Formation formation) {
        formations.add(formation);
    }

    public double getVolume() {
        return widthMeters * heightMeters * lengthMeters;
    }

    public String getName() {
        return name;
    }

    public double getWidthMeters() {
        return widthMeters;
    }

    public double getHeightMeters() {
        return heightMeters;
    }

    public double getLengthMeters() {
        return lengthMeters;
    }

    public List<Formation> getFormations() {
        return List.copyOf(formations);
    }

    @Override
    public String toString() {
        return String.format("Chamber[%s, %.1f x %.1f x %.1fm, formations=%d]",
                name, widthMeters, heightMeters, lengthMeters, formations.size());
    }
}
